README.md
# ssizeRNA
Sample size calculation for RNAseq data using limma/voom modeling assumptions

![sample size plot](demo/ssizeRNA_vary.png "Plot generated from example code")

```R
library(edgeR)
library(Biobase)
data(hammer.eset)
trt <- hammer.eset$protocol[ which(hammer.eset$Time == "2 weeks") ] 
counts <- exprs(hammer.eset)[, which(hammer.eset$Time == "2 weeks")]
counts <- counts[which(rowSums(counts) > 0), ]
geom.mean = function(row){
  row[row == 0] = 0.1
  if(length(row) != 0) return(exp(sum(log(row))/length(row)))
  else return(0)
}

mu <- apply(counts, 1, geom.mean)        ## geometric mean for each gene
d <- DGEList(counts)
d <- calcNormFactors(d)
d <- estimateCommonDisp(d)
d <- estimateTagwiseDisp(d)
disp <- d$tagwise.dispersion             ## dispersion for each gene

size <- ssizeRNA_vary(pi0=0.8, mu=mu, disp=disp, logfc=1, m=30, maxN=15, replace=F)
```


inst/NEWS.txt
CHANGES IN VERSION 1.3.2
------------------------

    o Update install information
    
CHANGES IN VERSION 1.3.1
------------------------

    o Changed email address
    

CHANGES IN VERSION 1.2.9
------------------------

    o Changed vignettes file
    
    
CHANGES IN VERSION 1.2.8
------------------------

    o Changed input logfc to fc


CHANGES IN VERSION 1.2.6
------------------------

    o Modified reference information


CHANGES IN VERSION 1.2.4
------------------------

    o Polished functions 
    o Added package vignettes 
    o Added data file "hammer.eset.RData"


CHANGES IN VERSION 1.1.2
------------------------

FIRST VERSION

    o Proposed a sample size calculation method for RNA-seq experiments


R/sim.counts.R
#' RNA-seq Count Data Simulation from Negative-Binomial Distribution
#' 
#' This function simulates count data from Negative-Binomial distribution
#' for two-sample RNA-seq experiments with given mean, dispersion 
#' and fold change. 
#' A count data matrix is generated.
#' 
#' @import MASS
#' 
#' @param nGenes total number of genes, the default value is \code{10000}.
#' @param pi0 proportion of non-differentially expressed genes, 
#'            the default value is \code{0.8}.
#' @param m sample size per treatment group.
#' @param mu a vector (or scalar) of mean counts in control group 
#'           from which to simulate.
#' @param disp a vector (or scalar) of dispersion parameter 
#'             from which to simulate.
#' @param fc a vector (or scalar, or a function that takes an integer n 
#'                     and generates a vector of length n)
#'           of fold change for differentially expressed (DE) genes.  
#' @param up proportion of up-regulated genes among all DE genes, 
#'           the default value is \code{0.5}.
#' @param replace sample with or without replacement from given parameters. 
#'                See Details for more information.
#' 
#' @details If the total number of genes \code{nGenes} is larger 
#'          than length of \code{mu} or \code{disp}, 
#'          \code{replace} always equals \code{TRUE}.
#' 
#' @return \item{counts}{RNA-seq count data matrix.}
#' @return \item{group}{treatment group vector.}
#' @return \item{lambda0}{mean counts in control group for each gene.}
#' @return \item{phi0}{dispersion parameter for each gene.}
#' @return \item{de}{differentially expressed genes indicator: 
#'                   \code{0} for non-differentially expressed genes, 
#'                   \code{1} for up-regulated genes, 
#'                   \code{-1} for down-regulated genes.}
#' @return \item{delta}{log2 fold change for each gene between 
#'                      treatment group and control group.}
#' 
#' @author Ran Bi \email{biranpier@@gmail.com}, 
#'         Peng Liu \email{pliu@@iastate.edu}
#' 
#' @examples
#' m <- 3                    ## sample size per treatment group
#' mu <- 10                  ## mean counts in control group for all genes 
#' disp <- 0.1               ## dispersion for all genes
#' fc <- 2                   ## 2-fold change for DE genes
#' 
#' sim <- sim.counts(m = m, mu = mu, disp = disp, fc = fc)
#' sim$counts                ## count data matrix
#' 
#' 
#' ## varying fold change
#' fc1 <- function(x){exp(rnorm(x, log(2), 0.5*log(2)))}
#' sim1 <- sim.counts(m = m, mu = mu, disp = disp, fc = fc1)
#' 
#' @export
#' 
sim.counts <- function(nGenes = 10000, pi0 = 0.8, m, mu, disp, fc, 
                       up = 0.5, replace = TRUE){
  arg <- list(nGenes = nGenes,
              pi0 = pi0,
              group = rep(c(1, 2), each = m))
 
  ## expected false positives
  FP <- round(nGenes * pi0)
  TP <- nGenes - FP 
  
  ## types of true positives
  TP_up <- round(TP * up)
  TP_down <- TP - TP_up 

  de <- c(rep(0, FP), rep(1, TP_up), rep(-1, TP_down))
  de <- de[sample.int(length(de))] ## resample
  
  # h = vector indicating which pseudo-genes to re-simulate
  h <- rep(TRUE, nGenes) 
  counts <- matrix(0, nrow = nGenes, ncol = 2 * m)
  
  ## log fold change, approximately half positive, half negative
  delta <- rep(0, nGenes)
  if (is.function(fc)){
    lfc <- log(fc(TP))
  }else{
    lfc <- log(fc)
  }
  delta[de != 0] <- lfc * de[de != 0]
  
  selected_genes <- true_means <- true_disps <- rep(0, nGenes)
  left_genes <- 1:length(mu)
  lambda <- phi <- matrix(0, nrow = nGenes, ncol = 2 * m)
  
  while(any(h)){
    temp <- sample.int(length(left_genes), sum(h), replace)
    temp <- temp[order(temp)]
    selected_genes[h] <- left_genes[temp]
    if (replace == FALSE){
      left_genes <- left_genes[-temp]
    }
    
    true_means[h] <- mu[selected_genes[h]]
    true_disps[h] <- disp[selected_genes[h]]
    
    lambda[h,] <- matrix(true_means[h], ncol = 1) %*% 
                  matrix(rep(1, 2 * m), nrow = 1) * 
                  cbind(matrix(rep(1, sum(h) * m), ncol = m), 
                        matrix(rep(exp(delta[h]), m), ncol = m))
    ## mean of counts
    
    phi[h,] <- matrix(rep(true_disps[h], 2 * m), ncol = 2 * m)
    ## dispersion of counts
    
    counts[h,] <- rnegbin(sum(h) * 2 * m, lambda[h,], 1 / phi[h,])
    h <- (rowSums(cpm(counts) > 2) < 3)
    # print(sum(h))
  }
  
  if(any(rowSums(cpm(counts) > 2) < 3 ))
    print("Error: Failed to simulate data: some genes are not expressed.")
  
  delta <- delta / log(2)
  
  list(counts = counts, 
       group = arg$group, 
       lambda0 = lambda[, 1],   # mean counts in control group
       phi0 = phi[, 1],   # dispersion
       de = de,   # DE indicator
       delta = delta  # log2 fold change
  )
}


R/check.power.R
#' Average Power and True FDR Based on limma/voom RNAseq Analysis Pipeline
#' 
#' For the limma/voom RNAseq analysis pipeline, when we control false discovery
#' rate by using the Benjamini and Hochberg step-up procedure (1995) 
#' and/or Storey and Tibshirani's q-value procedure (Storey et al, 2004),
#' \code{check.power} calculates average power and true FDR for given sample 
#' size, user-specified proportions of non-differentially expressed genes, 
#' number of iterations, FDR level to control, mean counts in control group, 
#' dispersion, and fold change.
#' 
#' @import qvalue stats
#'
#' @param nGenes total number of genes, the default value is \code{10000}.
#' @param pi0 proportion of non-differentially expressed genes, 
#'            the default value is \code{0.8}.
#' @param m sample size per treatment group.
#' @param mu a vector (or scalar) of mean counts in control group 
#'           from which to simulate.
#' @param disp a vector (or scalar) of dispersion parameter 
#'             from which to simulate.
#' @param fc a vector (or scalar, or a function that takes an integer n 
#'                     and generates a vector of length n)
#'           of fold change for differentially expressed (DE) genes.  
#' @param up proportion of up-regulated genes among all DE genes, 
#'           the default value is \code{0.5}.
#' @param replace sample with or without replacement from given parameters. 
#'                See Details for more information.
#' @param fdr the false discovery rate to be controlled.
#' @param sims number of simulations to run when computing power and FDR.
#' 
#' @return \item{pow_bh_ave}{average power when controlling FDR 
#'                           by Benjamini and Hochberg (1995) method.}
#' @return \item{fdr_bh_ave}{true false discovery rate when controlling FDR 
#'                           by Benjamini and Hochberg (1995) method.}
#' @return \item{pow_bh_ave}{average power when controlling FDR 
#'                           by q-value procedure (Storey et al., 2004).}
#' @return \item{fdr_bh_ave}{true false discovery rate when controlling FDR
#'                           by q-value procedure (Storey et al., 2004).}
#' 
#' @author Ran Bi \email{biranpier@@gmail.com}, 
#'         Peng Liu \email{pliu@@iastate.edu}
#' 
#' @references Benjamini, Y. and Hochberg, Y. (1995) 
#'             Controlling the false discovery rate: a practical and 
#'             powerful approach to multiple testing. 
#'             \emph{J. R. Stat. Soc. B}, 57, 289-300.
#' 
#'             Storey, J. D., Taylor, J. E. and Siegmund, D. (2004)
#'             Strong control, conservative point estimation and 
#'             simultaneous rates: a unified approach. 
#'             \emph{J. R. Stat. Soc. B}, 66, 187- 205.
#' 
#' @examples
#' library(limma)
#' library(qvalue)
#' m <- 14                      ## sample size per treatment group
#' mu <- 10                     ## mean read counts in control group
#' disp <- 0.1                  ## dispersion for all genes
#' fc <- 2                      ## 2-fold change for DE genes
#' 
#' check.power(m = m, mu = mu, disp = disp, fc = fc, sims = 2)
#'
#' @export
#' 
check.power <- function(nGenes = 10000, pi0 = 0.8, m, mu, disp, fc, 
                        up = 0.5, replace = TRUE, fdr = 0.05, sims = 100) {

  ## empirical "power" & "fdr" function
  powerfdr.fun <- function(fdr, p){
    V <- sum( (p < fdr) & (sim$de == FALSE))
    R <- sum( p < fdr )
    S <- R - V
    power <- S / (nGenes * (1 - pi0))
    fdr_true <- V / R
    return(c(power, fdr_true))
  }
  
  res <- list()
  pow_bh <- fdr_bh <- pow_qvalue <- fdr_qvalue <- rep(0, sims)
  for (j in 1:sims){
    # message("Performing simulation ", j, "/", sims, "...")
    sim <- sim.counts(nGenes, pi0, m, mu, disp, fc, up, replace)
    cts <- sim$counts
    lib.size <- colSums(cts)
    group = rep(c(1, 2), each = m)
    d <- DGEList(cts, lib.size, group = group)
    d <- calcNormFactors(d)
    design <- model.matrix(~ factor(group))
    y <- voom(d, design, plot=FALSE)       # convert read counts to log2-cpm 
                                           # with associated weights
    fit <- lmFit(y, design)
    fit <- eBayes(fit)
    pvalue <- fit$p.value[, 2]             # pvalue
    
    p_bh <- p.adjust(pvalue, method = "BH")  # Benjamini & Hochberg
    pow_bh[j] <- powerfdr.fun(fdr, p_bh)[1]
    fdr_bh[j] <- powerfdr.fun(fdr, p_bh)[2]
  
    p_qvalue <- qvalue(pvalue)$qvalues         # Storey and Tibshirani
    pow_qvalue[j] <- powerfdr.fun(fdr, p_qvalue)[1]
    fdr_qvalue[j] <- powerfdr.fun(fdr, p_qvalue)[2]
  }
  # message("Simulations completed.")
    
  ## average power & true fdr over sims simulations
  res$pow_bh_ave <- mean(pow_bh)
  res$fdr_bh_ave <- mean(fdr_bh)
  res$pow_qvalue_ave <- mean(pow_qvalue)
  res$fdr_qvalue_ave <- mean(fdr_qvalue)
  return(res)
}


R/ssizeRNA_single.R
#' Sample Size Calculations for Two-Sample RNA-seq Experiments 
#' with Single Set of Parameters
#' 
#' This function calculates appropriate sample sizes for two-sample 
#' RNA-seq experiments for a desired power in which mean and 
#' dispersion parameters are identical for all genes. 
#' Sample size calculations are performed at controlled false discovery rates, 
#' user-specified proportions of non-differentially expressed genes, 
#' mean counts in control group, dispersion, and fold change. 
#' A plot of power versus sample size is generated.
#' 
#' @import edgeR limma stats
#' 
#' @param nGenes total number of genes, the default value is \code{10000}.
#' @param pi0 proportion of non-differentially expressed genes, 
#'            the default value is \code{0.8}.
#' @param m pseudo sample size for generated data.
#' @param mu a vector (or scalar) of mean counts in control group 
#'           from which to simulate.
#' @param disp a vector (or scalar) of dispersion parameter 
#'             from which to simulate.
#' @param fc a vector (or scalar, or a function that takes an integer n 
#'                     and generates a vector of length n)
#'           of fold change for differentially expressed (DE) genes.  
#' @param up proportion of up-regulated genes among all DE genes, 
#'           the default value is \code{0.5}.
#' @param replace sample with or without replacement from given parameters. 
#'                See Details for more information.
#' @param fdr the false discovery rate to be controlled.
#' @param power the desired power to be achieved.
#' @param maxN the maximum sample size used for power calculations.
#' @param side options are "two-sided", "upper", or "lower".
#' @param cex.title controls size of chart titles.
#' @param cex.legend controls size of chart legend.
#' 
#' @details If a vector is input for \code{pi0}, sample size calculations 
#'          are performed for each proportion.
#' 
#' If the total number of genes is larger than length of \code{mu} or 
#' \code{disp}, \code{replace} always equals \code{TRUE}.
#' 
#' @return \item{ssize}{sample sizes (for each treatment) at which 
#'                      desired power is first reached.}
#' @return \item{power}{power calculations with corresponding sample sizes.}
#' @return \item{crit.vals}{critical value calculations with 
#'                          corresponding sample sizes.}
#' 
#' @author Ran Bi \email{biranpier@@gmail.com}, 
#'         Peng Liu \email{pliu@@iastate.edu}
#' 
#' @references Liu, P. and Hwang, J. T. G. (2007) Quick calculation for 
#' sample size while controlling false discovery rate with application 
#' to microarray analysis. \emph{Bioinformatics} 23(6): 739-746. 
#' 
#' Orr, M. and Liu, P. (2009) Sample size estimation while controlling 
#' false discovery rate for microarray experiments using ssize.fdr package. 
#' \emph{The R Journal}, 1, 1, May 2009, 47-53. 
#' 
#' Law, C. W., Chen, Y., Shi, W., Smyth, G. K. (2014). Voom: precision weights 
#' unlock linear model analysis tools for RNA-seq read counts. 
#' \emph{Genome Biology} 15, R29.
#' 
#' @seealso \code{\link{ssizeRNA_vary}}
#' 
#' @examples
#' mu <- 10                ## mean counts in control group for all genes
#' disp <- 0.1             ## dispersion for all genes
#' fc <- 2                 ## 2-fold change for DE genes
#' 
#' size <- ssizeRNA_single(m = 30, mu = mu, disp = disp, fc = fc, 
#'                         maxN = 20)
#' size$ssize              ## first sample size to reach desired power
#' size$power              ## calculated power for each sample size
#' size$crit.vals          ## calculated critical value for each sample size
#' 
#' @export
#' 
ssizeRNA_single <- function(nGenes = 10000, pi0 = 0.8, m = 200, mu, disp, 
                            fc, up = 0.5, replace = TRUE, fdr = 0.05, 
                            power = 0.8, maxN = 35, side = "two-sided", 
                            cex.title = 1.15, cex.legend = 1){ 
  
  sim <- sim.counts(nGenes, pi0, m, mu, disp, fc, up, replace)

  d_cpm <- DGEList(sim$counts)
  d_cpm <- calcNormFactors(d_cpm)
  group = rep(c(1, 2), each = m)  # treatment groups
  design <- model.matrix(~ factor(group))
  y <- voom(d_cpm, design, plot = FALSE)
  # convert read counts to log2-cpm with associated weights
  
  fit <- lmFit(y, design)
  fit <- eBayes(fit)
  fit$logcpm <- y$E  # normalized log-cpm value
  fit$weights <- y$weights  # precision weights for each observation 
  fit$www <- sqrt(2/m * apply(y$weights[, 1:m], 1, sum) * 
                    apply(y$weights[, (m+1):(2*m)], 1, sum) /
                    apply(y$weights, 1, sum))
  fit$Delta <- fit$coef[, 2] * fit$www  
  # effect size defined as weighted mean difference of log-cpm values, Delta_g
  
  dm <- mean((fit$Delta * sim$de)[sim$de!=0])
  ds <- sd((fit$Delta * sim$de)[sim$de!=0])
  sig <- density(fit$sigma)$x[which.max(density(fit$sigma)$y)]
  
  ret <- ssize.twoSampVaryDelta(deltaMean = dm, deltaSE = ds, sigma = sig, 
                                fdr = fdr, power = power, pi0 = pi0, 
                                maxN = maxN, side = side, 
                                cex.title = cex.title, cex.legend = cex.legend)
  return(ret)
}


R/ssize.twoSampVaryDelta.R
#' Sample Size Calculations for Two-Sample Microarray Experiments with 
#' Differing Mean Expressions but fixed Standard Deviations Among Genes
#' 
#' For given desired power, controlled false discovery rate, 
#' and user-specified proportions of non-differentially expressed genes, 
#' \code{ssize.twoSampVaryDelta} calculates appropriate sample sizes for 
#' two-sample microarray experiments in which the differences between mean 
#' treatment expression levels (\emph{delta.g} for gene \emph{g}) 
#' vary among genes. 
#' A plot of power versus sample size is generated.
#' 
#' @import stats graphics
#' 
#' @param deltaMean location (mean) parameter of normal distribution 
#'                  followed by each \emph{delta.g}.
#' @param deltaSE scale (standard deviation) parameter of normal distribution 
#'                followed by each \emph{delta.g}.
#' @param sigma the common standard deviation of expressions for all genes.
#' @param fdr the false discovery rate to be controlled.
#' @param power the desired power to be achieved.
#' @param pi0 a vector (or scalar) of proportions of non-differentially 
#'            expressed genes.
#' @param maxN the maximum sample size used for power calculations.
#' @param side options are "two-sided", "upper", or "lower".
#' @param cex.title controls size of chart titles.
#' @param cex.legend controls size of chart legend.
#' 
#' @details Each \emph{delta.g} is assumed to follow a Normal distribution 
#' with mean \code{deltaMean} and standard deviation \code{deltaSE}. 
#' The standard deviations of expressions are assumed identical for all genes.
#'
#' If a vector is input for \code{pi0}, sample size calculations 
#' are performed for each proportion.
#' 
#' @return \item{ssize}{sample sizes (for each treatment) at 
#'                      which desired power is first reached.}
#' @return \item{power}{power calculations with corresponding 
#'                      sample sizes.}
#' @return \item{crit.vals}{critical value calculations with 
#'                          corresponding sample sizes.}
#' 
#' @author Ran Bi \email{biranpier@@gmail.com}, 
#'         Peng Liu \email{pliu@@iastate.edu}
#' 
#' @references Liu, P. and Hwang, J. T. G. (2007) Quick calculation for 
#' sample size while controlling false discovery rate with application 
#' to microarray analysis. \emph{Bioinformatics} 23(6): 739-746. 
#' 
#' Orr, M. and Liu, P. (2009) Sample size estimation while controlling 
#' false discovery rate for microarray experiments using ssize.fdr package. 
#' \emph{The R Journal}, 1, 1, May 2009, 47-53. 
#' 
#' @seealso \code{\link{ssize.twoSamp}}, \code{\link{ssize.twoSampVary}}, 
#'          \code{\link{ssize.oneSamp}}, \code{\link{ssize.oneSampVary}}, 
#'          \code{\link{ssize.F}}, \code{\link{ssize.Fvary}}
#' 
#' @examples
#' dm <- 1.2; ds <- 0.1  ## the delta.g's follow a Normal(1.2, 0.1) distribution
#' s <- 1                ## common standard deviation
#' fdr <- 0.05           ## false discovery rate to be controlled
#' pwr <- 0.8            ## desired power
#' pi0 <- c(0.5, 0.8, 0.99) ## proportions of non-differentially expressed genes
#' N <- 35               ## maximum sample size for calculations
#' 
#' size <- ssize.twoSampVaryDelta(deltaMean = dm, deltaSE = ds, sigma = s, 
#'                                fdr = fdr, power = pwr, pi0 = pi0, 
#'                                maxN = N, side = "two-sided")
#' size$ssize                ## first sample size(s) to reach desired power
#' size$power                ## calculated power for each sample size
#' size$crit.vals            ## calculated critical value for each sample size
#' 
#' @export
#' 
ssize.twoSampVaryDelta <- function (deltaMean, deltaSE, sigma, fdr = 0.05, 
                                    power = 0.8, pi0 = 0.95, maxN = 35, 
                                    side = "two-sided", cex.title = 1.15, 
                                    cex.legend = 1){
  
  N <- maxN
  a <- fdr
  p <- pi0
  if (side == "two-sided") {
    TSVaryDelta <- function(c) {
      r <- a * (1 - p)/((1 - a) * p)
      dif <- abs((2 * pt(q = -c, df = 2 * n - 2)/
                    (1 - pt(q = c/sqrt(deltaSE^2 / (sigma^2 * 2/n) +1), 
                            df = 2 * n - 2, 
                            ncp = deltaMean/sqrt(deltaSE^2 + sigma^2 * 2/n))
                     + pt(q = -c/sqrt(deltaSE^2 / (sigma^2 * 2/n) +1), 
                          df = 2 * n - 2, 
                          ncp = deltaMean/sqrt(deltaSE^2 + sigma^2 * 2/n))) 
                  - r))
      return(dif)
    }
  }
  if (side == "upper") {
    TSVaryDelta <- function(c) {
      r <- a * (1 - p)/((1 - a) * p)
      dif <- abs((2 * pt(q = -c, df = 2 * n - 2)/
                    (1 - pt(q = c/sqrt(deltaSE^2 / (sigma^2 * 2/n) +1), 
                            df = 2 * n - 2, 
                            ncp = deltaMean/sqrt(deltaSE^2 + sigma^2 * 2/n)))
                  - r))
      return(dif)
    }
  }
  if (side == "lower") {
    TSVaryDelta <- function(c) {
      r <- a * (1 - p)/((1 - a) * p)
      dif <- abs((2 * pt(q = -c, df = 2 * n - 2)/
                    pt(q = -c/sqrt(deltaSE^2 / (sigma^2 * 2/n) +1), 
                       df = 2 * n - 2, 
                       ncp = deltaMean/sqrt(deltaSE^2 + sigma^2 * 2/n)) - r))
      return(dif)
    }
  }
  
  pwr2 <- NULL
  crit <- NULL
  ssize <- matrix(0, nrow = length(pi0), ncol = 3)
  colnames(ssize) <- c("pi0", "ssize", "power")
  up.start <- 50
  for (i in 1:length(pi0)) {
    p <- pi0[i]
    up <- up.start
    for (n in 3:N) {
      ci <- optimize(f = TSVaryDelta, interval = c(0, up))$min
      up <- ci
      if (abs(ci - up.start) >= 1) {
        if (side == "two-sided") {
          pwr.new <- (1 - pt(q = ci/sqrt(deltaSE^2 / (sigma^2 * 2/n) +1), 
                             df = 2 * n - 2, 
                             ncp = deltaMean/sqrt(deltaSE^2 + sigma^2 * 2/n))
                      + pt(q = -ci/sqrt(deltaSE^2 / (sigma^2 * 2/n) +1), 
                           df = 2 * n - 2, 
                           ncp = deltaMean/sqrt(deltaSE^2 + sigma^2 * 2/n)))
        }
      }
      if (abs(ci - up.start) < 1) {
        pwr.new <- 0
        ci <- NA
      }
      crit <- c(crit, ci)
      pwr2 <- c(pwr2, pwr.new)
      if (pwr2[(i - 1) * (N - 2) + n - 2] >= power & ssize[i, 1] == 0) {
        ssize[i, ] <- c(p, n, pwr2[(i - 1) * (N - 2) + n - 2])
      }
    }
  }
  ssize[, 1] <- pi0
  if (sum(ssize == 0) > 0) {
    warning("Desired power not achieved for at least one pi0")
  }
  ssize[ssize == 0] <- NA
  pwrMatrix <- matrix(c(3:N, pwr2), ncol = length(pi0) + 1, 
                      byrow = FALSE)
  for (i in 1:length(pi0)) {
    if (i == 1) {
      plot(3:N, pwrMatrix[, i + 1], col = i, xlim = c(0, N), ylim = c(0, 1), 
           xlab = "", ylab = "", pch = 16)
      lines(3:N, pwrMatrix[, i + 1], col = i, lty = i)
    }
    if (i != 1) {
      points(3:N, pwrMatrix[, i + 1], col = i, pch = 16)
      lines(3:N, pwrMatrix[, i + 1], col = i, lty = i)
    }
  }
  abline(h = power, lty = 2, lwd = 2)
  abline(v = 0:N, h = 0.1 * (0:10), col = "gray", lty = 3)
  title(xlab = "Sample size (n)", ylab = "Power")
  mtext(bquote(paste("Average power vs. sample size with fdr=", .(fdr), ",")), 
        cex = cex.title, padj = -1.85)
  mtext(bquote(paste(Delta[g], "~N(", .(round(deltaMean, 4)), ",", 
                     .(round(deltaSE, 4)), ") and ", sigma[g], " = ", 
                     .(round(sigma, 4)))), cex = cex.title, padj = -0.1)
  legend(x = N, y = 0, xjust = 1, yjust = 0, col = 1:i, pch = c(16, 16, 16), 
         lty = 1:length(pi0), legend = as.character(pi0), 
         bg = "white", title = expression(pi[0]), cex = cex.legend)
  pwrMatrix <- round(pwrMatrix, 7)
  colnames(pwrMatrix) <- c("n", as.character(pi0))
  critMatrix <- matrix(c(3:N, crit), ncol = length(pi0) + 1, byrow = FALSE)
  colnames(critMatrix) <- c("n", as.character(pi0))
  ret <- NULL
  ret$ssize <- ssize
  ret$power <- pwrMatrix
  ret$crit.vals <- critMatrix
  return(ret)
}

R/ssizeRNA_vary.R
#' Sample Size Calculations for Two-Sample RNA-seq Experiments with 
#' Differing Mean and Dispersion Among Genes
#' 
#' This function calculates appropriate sample sizes for two-sample 
#' RNA-seq experiments for a desired power in which mean and 
#' dispersion vary among genes. 
#' Sample size calculations are performed at controlled false discovery rates, 
#' user-specified proportions of non-differentially expressed genes, 
#' mean counts in control group, dispersion, and fold change. 
#' A plot of power versus sample size is generated.
#' 
#' @import Biobase ssize.fdr stats
#' 
#' @param nGenes total number of genes, the default value is \code{10000}.
#' @param pi0 proportion of non-differentially expressed genes, 
#'            the default value is \code{0.8}.
#' @param m pseudo sample size for generated data.
#' @param mu a vector (or scalar) of mean counts in control group 
#'           from which to simulate.
#' @param disp a vector (or scalar) of dispersion parameter 
#'             from which to simulate.
#' @param fc a vector (or scalar, or a function that takes an integer n 
#'                     and generates a vector of length n)
#'           of fold change for differentially expressed (DE) genes.  
#' @param up proportion of up-regulated genes among all DE genes, 
#'           the default value is \code{0.5}.
#' @param replace sample with or without replacement from given parameters. 
#'                See Details for more information.
#' @param fdr the false discovery rate to be controlled.
#' @param power the desired power to be achieved.
#' @param maxN the maximum sample size used for power calculations.
#' @param side options are "two-sided", "upper", or "lower".
#' @param cex.title controls size of chart titles.
#' @param cex.legend controls size of chart legend.
#' 
#' @details If a vector is input for \code{pi0}, sample size calculations 
#'          are performed for each proportion.
#' 
#' If the total number of genes is larger than length of \code{mu} or 
#' \code{disp}, \code{replace} always equals \code{TRUE}.
#' 
#' @return \item{ssize}{sample sizes (for each treatment) at which 
#'                      desired power is first reached.}
#' @return \item{power}{power calculations with corresponding sample sizes.}
#' @return \item{crit.vals}{critical value calculations with 
#'                          corresponding sample sizes.}
#' 
#' @author Ran Bi \email{biranpier@@gmail.com}, 
#'         Peng Liu \email{pliu@@iastate.edu}
#' 
#' @references Liu, P. and Hwang, J. T. G. (2007) Quick calculation for 
#' sample size while controlling false discovery rate with application 
#' to microarray analysis. \emph{Bioinformatics} 23(6): 739-746. 
#' 
#' Orr, M. and Liu, P. (2009) Sample size estimation while controlling 
#' false discovery rate for microarray experiments using ssize.fdr package. 
#' \emph{The R Journal}, 1, 1, May 2009, 47-53. 
#' 
#' Law, C. W., Chen, Y., Shi, W., Smyth, G. K. (2014). Voom: precision weights 
#' unlock linear model analysis tools for RNA-seq read counts. 
#' \emph{Genome Biology} 15, R29.
#' 
#' @seealso \code{\link{ssizeRNA_single}}
#' 
#' @examples
#' library(edgeR)
#' library(Biobase)
#' data(hammer.eset)
#' ## load hammer dataset (Hammer, P. et al., 2010)
#'
#' counts <- exprs(hammer.eset)[, phenoData(hammer.eset)$Time == "2 weeks"]
#' counts <- counts[rowSums(counts) > 0,]
#' trt <- hammer.eset$protocol[which(hammer.eset$Time == "2 weeks")] 
#' 
#' mu <- apply(counts[, trt == "control"], 1, mean)  
#' ## average read count in control group for each gene
#' 
#' d <- DGEList(counts)
#' d <- calcNormFactors(d)
#' d <- estimateCommonDisp(d)
#' d <- estimateTagwiseDisp(d)
#' disp <- d$tagwise.dispersion      
#' ## dispersion for each gene
#' 
#' ## fixed fold change
#' fc <- 2
#' size <- ssizeRNA_vary(mu = mu, disp = disp, fc = fc, 
#'                       m = 30, maxN = 15, replace = FALSE)
#' size$ssize         ## first sample size to reach desired power
#' size$power         ## calculated power for each sample size
#' size$crit.vals     ## calculated critical value for each sample size
#' 
#' 
#' ## varying fold change
#' ## fc1 <- function(x){exp(rnorm(x, log(2), 0.5*log(2)))}
#' ## size1 <- ssizeRNA_vary(pi0 = 0.8, mu = mu, disp = disp, fc = fc1, 
#' ##                        m = 30, maxN = 20, replace = FALSE)
#' 
#' @export
#' 
ssizeRNA_vary <- function(nGenes = 10000, pi0 = 0.8, m = 200, mu, disp, 
                          fc, up = 0.5, replace = TRUE, fdr = 0.05, 
                          power = 0.8, maxN = 35, side = "two-sided", 
                          cex.title = 1.15, cex.legend = 1) {  
  
  sim <- sim.counts(nGenes, pi0, m, mu, disp, fc, up, replace)
  
  d_cpm <- DGEList(sim$counts)
  d_cpm <- calcNormFactors(d_cpm)
  group = rep(c(1, 2), each = m)  # treatment groups
  design <- model.matrix(~factor(group))
  y <- voom(d_cpm, design, plot = FALSE)  
  # convert counts to log2-cpm w/weights
  
  fit <- lmFit(y, design)
  fit <- eBayes(fit)
  fit$logcpm <- y$E  # normalized log-cpm value
  fit$weights <- y$weights  # precision weights for each observation 

  from <- m + 1
  to <- 2 * m 
  fit$www <- sqrt(2 / m * apply(y$weights[, 1:m], 1, sum) * 
                  apply(y$weights[, from:to], 1, sum) / 
                  apply(y$weights, 1, sum))
  fit$Delta <- fit$coef[, 2] * fit$www  
  # effect == weighted mean diff(logcpm)
   
  dm <- mean( (fit$Delta * sim$de)[sim$de != 0] )
  ds <- sd( (fit$Delta * sim$de)[sim$de != 0] )
  a <- fit$df.prior / 2
  b <- fit$df.prior * fit$s2.prior / 2
  sig <- density(fit$sigma)$x[which.max(density(fit$sigma)$y)]
  
  ret <- ssize.twoSampVary(deltaMean = dm, deltaSE = ds, a = a, b = b, 
                           fdr = fdr, power = power, pi0 = pi0, 
                           maxN = maxN, side = side, cex.title = cex.title, 
                           cex.legend = cex.legend)
  return(ret)
}
